File: code/2-AddTwoNumbers/cpp/solution2.cpp
#include <iostream>
using namespace std;

class Solution2{
    ListNode* Solution(ListNode* l1, ListNode* l2, int carry) {
        if( !l1 && !l2 && !carry)
            return nullptr;
        
        int value = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
        carry = value / 10;
        ListNode* ptr = new ListNode(value % 10);
        ptr->next = Solution(l1 ? l1->next : nullptr, l2 ? l2->next : nullptr, carry);
        return ptr;
    }

    ListNode* Solution(ListNode* l1, ListNode* l2) {
        return Solution(l1, l2, 0);
    }
}

File: code/No2/java/Solution2.java
/**
 * @author Patrick-lenovo
 * @create 2018/6/16 21:57
 **/
public class Solution2 {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addTwoNumbers(l1, l2, 0);
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2, int carry) {

        if (l1 == null && l2 == null && carry == 0) {
            return null;
        }

        int val = (l1 == null ? 0 : l1.val) + (l2 == null ? 0 : l2.val) + carry;
        ListNode p = new ListNode(val % 10);
        p.next = addTwoNumbers(l1 == null ? l1 : l1.next, l2 == null ? l2 : l2.next, val / 10);
        return p;
    }
}
